### DISCUSSION TITLE
Clearer Custom Error Page documentation page

### DISCUSSION THREAD

**Discussion by leo-petrucci** [vercel/next.js#17778] [Help] [answered]
# Bug report

## Describe the bug

I'm unsure if this can be counted as a bug, but I think the documentation page is missing information, or the example isn't actually working properly. I'm more than happy to submit a PR for changing the documentation page, but I wanted to make sure if I'm wrong before doing so.

The problem lies with the [Custom Error Page](https://nextjs.org/docs/advanced-features/custom-error-page) specifically with the third example.

The docs are telling people that to display the error page yourself you simply need to return it from main component:
```
export default function Page({ errorCode, stars }) {
  if (errorCode) {
    return <Error statusCode={errorCode} />
  }

  return <div>Next stars: {stars}</div>
}
```

And while this will render the error page, it actually won't set the correct status code, you can check this in the network tab of the your browser's dev tools, the response will always be **200**.

To return the correct error code I need to do so in `getServerSideProps`:
```
export async function getServerSideProps({response}) {
  const res = await fetch('https://api.github.com/repos/vercel/next.js')
  const errorCode = res.ok ? false : res.statusCode
  const json = await res.json()

  // This line is missing from the example
  response.statusCode = errorCode 

  return {
    props: { errorCode, stars: json.stargazers_count },
  }
}
```

I was very confused why Google kept indexing pages that were returning 404 for me before I realised this was the problem so I think it's worth adding to the documentation to avoid any further confusion. Especially since this is so important for SEO.

Let me know if this is worth changing, maybe I'm the only one that thinks this is useful 😂

**Comment by alex-grover** [comment]
I don't think there's an error in the documentation. You shouldn't be able to see a network request in your dev tools because the `fetch` is only happening on the server. What you're doing is causing the next server to think that an exception happened when it actually didn't.

The confusion might be happening because the `fetch('https://api.github.com/repos/vercel/next.js')` will likely 200 - that's a valid API call, unless Github is having an outage. So no `errorCode` is actually being passed to your page component. If you change the fetch to point to a URL that returns a 4xx or 5xx you'll see the correct error page rendered.

You can simulate this and verify that the right thing is actually happening by adding a `return { errorCode: 400/422/500 }` in the gSSP function.

**Reply by leo-petrucci** [reply]
If a page you visit is a 404 you will get a 404 network request for the page you tried to visit. That's what google looks for when understanding if a page exists or not.

Go ahead and try it, even by forcing the GitHub request to 404 the page will still come back as a 200 and google won't index it correctly.

**Reply by alex-grover** [reply]
Got it, I misunderstood, I thought you were saying the UI of the page always showed an incorrect code. Confusion over which status code you were referring to!

**Reply by leo-petrucci** [reply]
@ajgrover Oops, probably my bad, english isn't my first language 😊

**Comment by lfades** [comment, marked as answer]
@creativiii That example is there to show you how to use the Error component inside another component and display an error, but it can't change the `response` object or the status code inside of it (it's not possible because it renders after the response was sent).

A RFC for custom status codes inside `getServerSideProps` may come in the future to handle this case (https://github.com/vercel/next.js/discussions/14890#discussioncomment-48497)

**Reply by leo-petrucci** [reply]
That's exactly the problem I'm talking about, surely the example to return error pages should also tell you to set the correct error code for the page?

It doesn't really matter if that's done in the component or during the request, but it should be done at **some** point.

I don't understand how one excludes the other.

**Reply by leerob** [reply]
@creativiii Please do not unmark an answer when selected by a maintainer. @lfades's answer is correct and the RFC provides more information for your comment.

**Reply by leerob** [reply]
If you have any further questions about this, please add details for what you'd like the implementation to look like on the RFC 🙏

**Reply by leo-petrucci** [reply]
@leerob Sorry! Didn't realise. 

What kind of other information should I provide? Surely returning an error component isn't very useful without actually making the page return a proper error?

At the moment the page is essentially showing that you can render a component rather than another, which doesn't seem very useful as that's fairly basic information about react. Even though you're rendering a component that says 404, the page is still returning a 200.

The real information the page should contain should be about telling search engines that the page they've visited is returning something else than 200, so they can handle it correctly rather than indexing an empty page 😅

**Reply by lfades** [reply]
@creativiii What you're currently asking for can't be resolved using the error page, none of the components are able to access the request or response objects, so changing the status code has to be done in the data fetching methods, and we will eventually add built-in support for that. You can already do redirects (it's an experimental feature) if the page is a 404.

**Reply by leo-petrucci** [reply]
@lfades Again I think you are misunderstanding what I'm saying.

I am aware you cannot change the response from the component, I'm not asking to do that.

 What I am saying is that the fact that the component doesn't actually change the response of the page isn't very clear from the documentation and it would probably be beneficial for all users to explain **how** to return an error in `getInitialProps`.

This may seem obvious to someone who is familiar with Next.js, but an average user will not notice the page hasn't actually returned a 404 without checking their network tab, so more should be done to explain how to actually return those errors.

**Reply by leo-petrucci** [reply]
To be clear, here's me and a bunch of other people being confused by how to actually get the page to 404 beyond showing an error component https://github.com/vercel/next.js/issues/11644

**Reply by leo-petrucci** [reply]
Any word on this? I really do think it's worth clarifying.
